[#LIBCLOUD-281] Support Gandi DNS

Jc2k opened the pull request:
This pull request adds a new DNS Driver for Gandi. It includes tests. It is pep8 compliant and also brings `libcloud/common/gandi.py` into pep8 compliance.

Note that whilst Gandi's API does expose a record id field it changes for each version of the zone (effectively each 'transaction' renumbers all record id's that libcloud sees). So we use the same approach as Route53 and use type:name as the record id.

Jc2k committed:
Initial Gandi DNS driver

Jc2k committed:
Test scaffolding for GandDNSDriver

Jc2k committed:
All positive tests pass

Jc2k committed:
Raise an exception when deleting a record that does not exist

Jc2k committed:
More passing tests

Jc2k committed:
Fixtures and more fixing

Commit:
Fixtures for version stuff

Commit:
Fixes for update_record and delete_record

Commit:
PEP8 fixery

Jc2k committed:
Add and test update_zone method

Jc2k committed:
Refactor version handling into a with statement

Jc2k committed:
Validate ttl is in range and that data payload isnt too large

Jc2k committed:
Hide gandi record id from users as its not especially useful

Jc2k committed:
Final checks, fixes and pep8 cleanup

Kami commented on libcloud/common/gandi.py:
```diff
@@ -93,7 +93,10 @@ def request(self, method, *args):
-            raise GandiException(1001, e)
+            self.parse_error(e.faultCode, e.faultString)
+
+    def parse_error(self, code, message):
```
Hm, you pass in `code`, but you never use it.

Method name is also slightly misleading to me - it actually throws an exception and doesn't do any parsing.

Jc2k replied:
Intention is for it to look like a normal Response class. It should be subclassed (and is, in the DNS case). The `code` attribute is used in the DNS driver. I was mightily annoyed to raise GandiException 1001, e but thats what the existing compute driver expects :(

Kami replied:
Ah, I see...

Would be nice to at least add a docstring which explains this then.

Kami commented on libcloud/dns/drivers/gandi.py:
```diff
@@ -0,0 +1,249 @@
+TTL_MAX = 2592000  # 30 days
+
+
+class NewZoneVersion(object):
```
Would be nice to add a short docstring which describes what this context manager does.

Kami commented on libcloud/dns/drivers/gandi.py:
```diff
@@ -0,0 +1,249 @@
+    def __exit__(self, type, value, traceback):
+        if not traceback:
+            zid = int(self.zone.id)
+            c = self.connection
```
Minor code style thing: We have an (atm) unwritten rule about the shortest variable name being 3 characters long

Kami commented on libcloud/dns/drivers/gandi.py:
```diff
@@ -0,0 +1,249 @@
+class GandiDNSConnection(GandiConnection):
+
+    def parse_error(self, code, message):
+        if code == 581042:
```
Doesn't have en else, is this expected? Or should we throw an unexpected code error if we don't recognize the code?

Jc2k replied:
The base class calls parse_error and is meant to throw an exception if the subclass doesn't throw a more specific one. This emulated the behaviour of parse_error in the normal connection/response objects.

It looks like I didn't get that right in the base class though!

Kami replied:
Yeah, I'm still slightly confused though because you never actually call a parent class method and you override it so the parent method will never end up being called in this case.

Jc2k replied:
Yeah, what i should have done is put the raise GandiExeption in the method that called parse_error, like is done in a normal response object :-(

Kami commented on libcloud/dns/drivers/gandi.py:
```diff
@@ -0,0 +1,249 @@
+class GandiDNSDriver(BaseGandiDriver, DNSDriver):
+    type = Provider.GANDI
+    name = 'Gandi DNS'
+    website = 'http://doc.rpc.gandi.net/domain/reference.html'
```
`website` attribute should point to the product page, not the docs.

Feel free to include link to the docs at the module level docstring though.

Kami commented on libcloud/dns/types.py:
```diff
@@ -57,7 +58,9 @@ class RecordType(object):
     GEO = 14
     URL = 15
-
+    WKS = 16
```
Heh, I actually had to google what `WKS` record type is since I never heard about it before http://tools.ietf.org/html/rfc1123#page-55.

It sounds like it's a deprecated record type, does it make sense to still support it here?

Jc2k replied:
It might exist in a Gandi zone as they are exposed by its API. It seems a shame for libcloud to not work on that zone for the sake of a single enum.

Kami replied:
Yeah, I'm in this case I'm fine with supporting it. Supporting it doesn't really make other code any harder to maintain for us.

Kami commented on libcloud/test/dns/test_gandi.py:
```diff
@@ -0,0 +1,287 @@
+    mockCls = GandiMockHttp
+
+
+class GandiTests(BaseGandiTests):
```
Minor style thing - common convention is to move test class at the top and MockHttp class at the bottom.

Jc2k replied:
I thought you might say that :)

GandiTests needs to point at a DummyTransport with a class attribute, which needs to point at a mock. Class attributes need to point at things that are defined, so this would explode.

I can do this other ways but they are ugly. My next patch for the 2 gandi drivers is to use the xmlrpc parsing but atop the libcloud HTTP stack. This will fix the some other issues we know the Gandi drivers have but also make the tests a bit more like the other tests, and I can fix this then. Or I can find some other way of setting the transport class up now?

Jc2k committed:
Fix comments from review

Jc2k committed:
Formatting tweak

Jc2k commented:
I've added the comments, nerfed the 1 character variables, and the parse_error thing does as intended.

I also noticed there was an unused import and removed it.

I'm started to investigate a follow up piece of work which will as a consequence fix the test/mock order, but in the mean time this is fully functional, the tests are passing (including an out-of-tree check i run against a real account) as is pep8 :-)

Kami commented:
@Jc2K Looks good.

Please attach a final patch to the jira ticket and I will perform a final review and merge it into trunk.

Jc2k commented:
Merged.
